### PR TITLE
SCIM: Optimize user queries: Use WITH and JOIN rather than subqueries

### PR DESCRIPTION
**Pull request by vdavid**
- Set out to solve https://github.com/sourcegraph/sourcegraph/issues/48899, as per @pjlast's [suggestion](https://github.com/sourcegraph/sourcegraph/pull/48816/files#r1129187068)

Just adding a "LEFT JOIN" doesn't work because a field called `deleted_at` exists both in `users` and `user_external_accounts`, and they clash after JOINing.
Aliasing doesn't work either, as discussed [here](https://github.com/sourcegraph/sourcegraph/pull/48291#discussion_r1119175162).
To solve it, I used @chwarwick's solution (see the same discussion linked above) and used a `WITH` subquery.

This _works_ 🚀 , BUT the new problem is the `EXPLAIN ANALYZE` result I got when I ran it on dotcom (based on [this guide](https://handbook.sourcegraph.com/departments/engineering/dev/process/deployments/postgresql/#sourcegraphcom-specific)):

My original query that included a subquery, plus the `EXPLAIN ANALYZE` results:

```
EXPLAIN ANALYZE SELECT u.id,
       u.username,
       u.display_name,
       u.avatar_url,
       u.created_at,
       u.updated_at,
       u.site_admin,
       u.passwd IS NOT NULL,
       u.tags,
       u.invalidated_sessions_at,
       u.tos_accepted,
       u.searchable,
       (SELECT COUNT(user_id) FROM user_external_accounts WHERE user_id=u.id AND service_type = 'scim') >= 1 AS scim_controlled
FROM users u;

Seq Scan on users u  (cost=0.00..308122.63 rows=128233 width=119) (actual time=0.015..90.713 rows=53176 loops=1)
   SubPlan 1
     ->  Aggregate  (cost=2.35..2.35 rows=1 width=8) (actual time=0.001..0.001 rows=1 loops=53176)
           ->  Index Only Scan using user_external_accounts_user_id_scim_service_type on user_external_accounts  (cost=0.12..2.34 rows=1 width=4) (actual time=0.000..0.000 rows=0 loops=53176)
                 Index Cond: (user_id = u.id)
                 Heap Fetches: 0
 Planning Time: 0.126 ms
 Execution Time: 93.453 ms
(8 rows)
```

The new one that uses `WITH` and `JOIN`, plus the `EXPLAIN ANALYZE` results:

```
EXPLAIN ANALYZE WITH scim_accounts AS (SELECT user_id FROM user_external_accounts WHERE service_type = 'scim')
SELECT u.id,
       u.username,
       u.display_name,
       u.avatar_url,
       u.created_at,
       u.updated_at,
       u.site_admin,
       u.passwd IS NOT NULL,
       u.tags,
       u.invalidated_sessions_at,
       u.tos_accepted,
       u.searchable,
       (SELECT COUNT(user_id) FROM scim_accounts WHERE user_id=u.id) >= 1 AS scim_controlled
FROM users u
LEFT JOIN scim_accounts sa ON u.id = sa.user_id;

Hash Left Join  (cost=2.37..11105.32 rows=128233 width=119) (actual time=0.345..87.601 rows=53176 loops=1)
   Hash Cond: (u.id = sa.user_id)
   CTE scim_accounts
     ->  Index Only Scan using user_external_accounts_user_id_scim_service_type on user_external_accounts  (cost=0.12..2.34 rows=1 width=4) (actual time=0.321..0.321 rows=0 loops=1)
           Heap Fetches: 0
   ->  Seq Scan on users u  (cost=0.00..5813.33 rows=128233 width=178) (actual time=0.008..28.870 rows=53176 loops=1)
   ->  Hash  (cost=0.02..0.02 rows=1 width=4) (actual time=0.322..0.322 rows=0 loops=1)
         Buckets: 1024  Batches: 1  Memory Usage: 8kB
         ->  CTE Scan on scim_accounts sa  (cost=0.00..0.02 rows=1 width=4) (actual time=0.321..0.322 rows=0 loops=1)
   SubPlan 2
     ->  Aggregate  (cost=0.02..0.03 rows=1 width=8) (actual time=0.000..0.000 rows=1 loops=53176)
           ->  CTE Scan on scim_accounts  (cost=0.00..0.02 rows=1 width=4) (actual time=0.000..0.000 rows=0 loops=53176)
                 Filter: (user_id = u.id)
 Planning Time: 0.183 ms
 Execution Time: 90.410 ms
(15 rows)
```

The time difference is smaller than what I'd expect, plus the seq scans in both cases make me worry. Am I looking at it wrong? → ✅  solved in comments.

Note on the priority of this PR: It's not a dealbreaker to have the subqueries there, but using a JOIN would be just much nicer.

## Test plan

- Tested it for functionality, it works!
- Tested that both queries use indexes all around.

### CODE REVIEW

File: internal/database/users.go [hunk 1]
```diff
@@ -1192,6 +1192,7 @@ func (u *userStore) getOneBySQL(ctx context.Context, q *sqlf.Query) (*types.User
 // getBySQL returns users matching the SQL query, if any exist.
 func (u *userStore) getBySQL(ctx context.Context, query *sqlf.Query) ([]*types.User, error) {
 	q := sqlf.Sprintf(`
+WITH scim_accounts AS (SELECT user_id FROM user_external_accounts WHERE service_type = 'scim')
```
Comment (eseliger) [inline internal/database/users.go:1195]: Found an alternative way that seems to be faster:

```
sg=> EXPLAIN ANALYZE
sg-> SELECT u.id,
sg->        u.username,
sg->        u.display_name,
sg->        u.avatar_url,
sg->        u.created_at,
sg->        u.updated_at,
sg->        u.site_admin,
sg->        u.passwd IS NOT NULL,
sg->        u.tags,
sg->        u.invalidated_sessions_at,
sg->        u.tos_accepted,
sg->        u.searchable,
sg->        EXISTS (
sg(>         SELECT 1 FROM user_external_accounts WHERE service_type = 'scim' AND user_id = u.id
sg(>        ) AS scim_controlled
sg-> FROM users u;
                                                                                                  QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Seq Scan on users u  (cost=0.00..306199.13 rows=128233 width=119) (actual time=0.012..28.053 rows=53176 loops=1)
   SubPlan 1
     ->  Index Only Scan using user_external_accounts_user_id_scim_service_type on user_external_accounts  (cost=0.12..2.34 rows=1 width=0) (never executed)
           Index Cond: (user_id = u.id)
           Heap Fetches: 0
   SubPlan 2
     ->  Index Only Scan using user_external_accounts_user_id_scim_service_type on user_external_accounts user_external_accounts_1  (cost=0.12..2.34 rows=1 width=4) (actual time=0.003..0.003 rows=0 loops=1)
           Heap Fetches: 0
 Planning Time: 0.151 ms
 Execution Time: 30.802 ms
(10 rows)
```

Comment (eseliger) [reply internal/database/users.go:1195]: 30ms seems reasonable, given this is listing ALL users on the instance, with pagination it'd be much faster:

```
sg=> EXPLAIN ANALYZE
sg-> SELECT u.id,
sg->        u.username,
sg->        u.display_name,
sg->        u.avatar_url,
sg->        u.created_at,
sg->        u.updated_at,
sg->        u.site_admin,
sg->        u.passwd IS NOT NULL,
sg->        u.tags,
sg->        u.invalidated_sessions_at,
sg->        u.tos_accepted,
sg->        u.searchable,
sg->        EXISTS (
sg(>         SELECT 1 FROM user_external_accounts WHERE service_type = 'scim' AND user_id = u.id
sg(>        ) AS scim_controlled
sg-> FROM users u
sg-> ORDER BY id ASC
sg-> LIMIT 50;
                                                                                                     QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.42..120.50 rows=50 width=119) (actual time=0.021..0.138 rows=50 loops=1)
   ->  Index Scan using users_pkey on users u  (cost=0.42..307969.25 rows=128233 width=119) (actual time=0.020..0.129 rows=50 loops=1)
         SubPlan 1
           ->  Index Only Scan using user_external_accounts_user_id_scim_service_type on user_external_accounts  (cost=0.12..2.34 rows=1 width=0) (never executed)
                 Index Cond: (user_id = u.id)
                 Heap Fetches: 0
         SubPlan 2
           ->  Index Only Scan using user_external_accounts_user_id_scim_service_type on user_external_accounts user_external_accounts_1  (cost=0.12..2.34 rows=1 width=4) (actual time=0.003..0.003 rows=0 loops=1)
                 Heap Fetches: 0
 Planning Time: 0.269 ms
 Execution Time: 0.197 ms
(11 rows)
```

Comment (chwarwick) [inline internal/database/users.go:1195]: You would want to apply the same (or similar) where clause here to limit by userID, the effect then would be like having a little cache for the subqueries.  

I wouldn't expect you would see much of a difference unless it was replacing multiple correlated subqueries

File: internal/database/users.go [hunk 2]
```diff
@@ -1204,8 +1205,10 @@ SELECT u.id,
        u.invalidated_sessions_at,
        u.tos_accepted,
        u.searchable,
-       (SELECT COUNT(user_id) FROM user_external_accounts WHERE user_id=u.id AND service_type = 'scim') >= 1 AS scim_controlled
-FROM users u %s`, query)
+       (SELECT COUNT(user_id) FROM scim_accounts WHERE user_id=u.id) >= 1 AS scim_controlled
+FROM users u
+LEFT JOIN scim_accounts sa ON u.id = sa.user_id
```
Comment (pjlast) [inline internal/database/users.go:1210]: Why can't this just be the following
```suggestion
       COUNT(uea.service_type) >= 1 AS scim_controlled
FROM users u
LEFT JOIN user_external_accounts uea
    ON u.id = uea.user_id
    AND uea.service_type = 'scim'
    AND uea.deleted_at IS NOT NULL
```

Comment (pjlast) [reply internal/database/users.go:1210]: And then group by

Comment (pjlast) [reply internal/database/users.go:1210]: But interestingly enough, this actually seems slower. The scim index makes the original one faster

Comment (pjlast) [reply internal/database/users.go:1210]: Without the index, the other query takes a massive amount of time. Try substituting `'scim'` with `'github'` and see how long it takes. But the one here finishes really fast even without a dedicated index 🤷‍♂️

Comment (pjlast) [reply internal/database/users.go:1210]: But yeah @eseliger 's solution is the fastest by far, I'd go with that, even when you add the `AND deleted_at IS NULL`

**Comment by vdavid** [comment]
Thank you for all your contributions, everyone! I went with @eseliger's solution for the first query and incorporated @pjlast's thought about filtering for non-deleted external accounts. Then I ran the `EXPLAIN ANALYZE` command a few more times to ensure that the final state of both queries is performant (it is). Now I'm confident to merge this.

**Comment by vdavid** [comment]
...if someone approves this. Please give this another round of review.

**Comment by eseliger** [comment]
did you actually go with my solution? 🤔 I don't see it

File: internal/database/users.go [hunk 3]
```diff
@@ -1232,6 +1234,15 @@ FROM users u %s`, query)
...
 const userForSCIMQueryFmtStr = `
+WITH scim_accounts AS (
+    SELECT
+        user_id,
+        account_id,
+        account_data
+    FROM user_external_accounts
+    WHERE service_type = 'scim'
+      AND deleted_at IS NULL
+)
```
Comment (eseliger) [inline internal/database/users.go:1245]: inconsistent indentation
```suggestion
        account_data
    FROM user_external_accounts
    WHERE service_type = 'scim'
        AND deleted_at IS NULL
)
```

**Review by eseliger — APPROVED** [review]

**Pull request merged**